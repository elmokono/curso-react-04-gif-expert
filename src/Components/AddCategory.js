import React, { useState } from 'react'
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    //change text
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
        console.log('- handleInputChange');
    }

    //callback parent component
    const handleSubmit = (e) => {
        e.preventDefault();
        console.log('- handleSubmit', inputValue);

        if (inputValue.trim().length < 3)
            return;

        setCategories(x => [inputValue.trim(), ...x]);
        setInputValue('');
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input placeholder='Input Category Name' type="text" value={inputValue} onChange={handleInputChange}></input>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};