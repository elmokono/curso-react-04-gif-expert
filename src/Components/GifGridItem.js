import PropTypes from "prop-types";

export const GifGridItem = ({ id, title, url }) => {
    return (
        <div className="card animate__animated animate__fadeIn">
            <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    );
}

GifGridItem.propTypes = {
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};

//UNIT TESTING
/*
configurar enzime
configurar enzime to json
test: debe de mostrar el componente correctamente
    -usar shallow
    -el wrapper debe matchear el snapshot
*/