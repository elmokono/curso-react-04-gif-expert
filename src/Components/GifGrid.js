import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from "../Hooks/useFetchGifs"

export const GifGrid = ({ category }) => {

    const { loading, data: images } = useFetchGifs(category); //custom hook

    return (
        <div>
            <h3>{category}</h3>
            { loading && <p className='animate__flash'>Loading</p> }
            <div className='card-grid'>
                {images.map((img) => {
                    return (<GifGridItem key={img.id} {...img} />)
                })}

            </div>
        </div>
    )
}
