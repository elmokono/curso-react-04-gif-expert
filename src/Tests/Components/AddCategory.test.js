import { shallow } from "enzyme"
import '@testing-library/jest-dom'
import { AddCategory } from "../../Components/AddCategory"

describe('test over addCategory', () => {

    const setCategoriesDummy = jest.fn(); //trackea llamadas
    let wrapper;//shallow(<AddCategory setCategories={setCategoriesDummy} />);

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategoriesDummy} />);
    });

    test('should display correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should invoke handleInput', () => {
        const textInput = wrapper.find('input');
        const textDummy = 'hola';
        textInput.simulate('change', { target: { value: textDummy } });

        const pValue = wrapper.find('p');
        expect(pValue.text().trim()).toBe(textDummy);
    })

    test('should NOT submit on short text', () => {
        wrapper.find('form').simulate('submit', { preventDefault: () => { } });
        expect(setCategoriesDummy).not.toHaveBeenCalled();
    })

    test('should clear input text after submit', () => {
        wrapper.find('input').simulate('change', { target: { value: 'hola mundo' } });
        wrapper.find('form').simulate('submit', { preventDefault: () => { } });
        expect(setCategoriesDummy).toHaveBeenCalledWith(expect.any(Function));
        expect(wrapper.find('input').prop('value')).toBe('');
    })
})