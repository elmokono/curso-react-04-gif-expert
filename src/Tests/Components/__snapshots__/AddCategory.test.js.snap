// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test over addCategory should display correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Input Category Name"
    type="text"
    value=""
  />
</form>
`;
